x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,391 @@
+	setup := Setup(t)
+
+	trusteeAccAddress1 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress2 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress3 := sdktestdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	for _, role := range []dclauthtypes.AccountRole{
+		dclauthtypes.TestHouse,
+		dclauthtypes.CertificationCenter,
+		dclauthtypes.Vendor,
+		dclauthtypes.NodeAdmin,
+	} {
+		accAddress := sdktestdata.GenerateAccAddress()
+		setup.AddAccount(accAddress, []dclauthtypes.AccountRole{role})
+
+		// propose upgrade by user without Trustee role
+		msgProposeUpgrade := NewMsgProposeUpgrade(accAddress)
+		_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+		require.Error(t, err)
+		require.True(t, sdkerrors.ErrUnauthorized.Is(err))
+		setup.UpgradeKeeper.AssertNotCalled(t, "ScheduleUpgrade", mock.Anything, msgProposeUpgrade.Plan)
+	}
+
+	// propose and approve upgrade by Trustees (3 Trustees, >=2/3 approvals needed)
+
+	// propose upgrade by user with Trustee role
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.NoError(t, err)
+
+	// approve upgrade
+	msgApproveUpgrade := NewMsgApproveUpgrade(trusteeAccAddress2)
+	_, err = setup.Handler(setup.Ctx, msgApproveUpgrade)
+	require.NoError(t, err)
+
+	setup.UpgradeKeeper.AssertCalled(
+		t,
+		"ScheduleUpgrade",
+		mock.MatchedBy(isContextWithNonCachedMultiStore),
+		msgProposeUpgrade.Plan,
+	)
+}
+
+func TestHandler_ApproveUpgrade(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress2 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress3 := sdktestdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	// propose and approve upgrade by Trustees (3 Trustees, >=2/3 approvals needed)
+
+	// propose new upgrade
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.NoError(t, err)
+
+	// approve upgrade
+	msgApproveUpgrade := NewMsgApproveUpgrade(trusteeAccAddress2)
+	_, err = setup.Handler(setup.Ctx, msgApproveUpgrade)
+	require.NoError(t, err)
+
+	setup.UpgradeKeeper.AssertCalled(
+		t,
+		"ScheduleUpgrade",
+		mock.MatchedBy(isContextWithNonCachedMultiStore),
+		msgProposeUpgrade.Plan,
+	)
```

Comment:
Also it must be verified that `ProposedUpgrade` does not exist anymore.

Reply:
Added